### PR TITLE
[Enhancement] Read pos-delete file using datacache

### PR DESCRIPTION
**Pull request by stephen-shelby**
## What I'm doing:
1. using sharedIOBuffer and cachedInputStream to access pos-delete files
2. add io metrics for reading delete files
3. refactor iceberg_delete_builder

Due to the significant optimization of pos-delete in the iceberg v3 spec, we temporarily use data cache to read pos-delete for optimization. 
https://docs.google.com/document/d/18Bqhr-vnzFfQk1S4AgRISkA_5_m5m32Nnc2Cw0zn2XM/edit?tab=t.0
https://github.com/apache/iceberg/pull/11302
https://github.com/apache/iceberg/pull/11238#discussion_r1798622067

Fixes #issue

https://github.com/StarRocks/starrocks/pull/51050

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Review by dirtysalt — DISMISSED** [review]

File: be/src/exec/iceberg/iceberg_delete_builder.cpp [hunk 1]
```diff
@@ -37,63 +38,147 @@ static const IcebergColumnMeta k_delete_file_path{
...
+    const OpenFileOptions options{.fs = _params.fs,
+                                  .path = delete_file.full_path,
+                                  .file_size = delete_file.length,
+                                  .fs_stats = &fs_scan_stats,
+                                  .app_stats = &app_scan_stats,
+                                  .datacache_options = _params.datacache_options};
+    ASSIGN_OR_RETURN(auto file,
+                     HdfsScanner::create_random_access_file(shared_buffered_input_stream, cache_input_stream, options));
+    std::vector<io::SharedBufferedInputStream::IORange> io_ranges{};
+    for (int64_t offset = 0; offset < delete_file.length;) {
```
Comment (zombee0) [inline be/src/exec/iceberg/iceberg_delete_builder.cpp:54]: while statement?

File: be/src/exec/iceberg/iceberg_delete_builder.cpp [hunk 2]
```diff
@@ -37,63 +38,147 @@ static const IcebergColumnMeta k_delete_file_path{
...
         }
     }
+    return Status::OK();
+}
+
+Status IcebergDeleteBuilder::build_parquet(const TIcebergDeleteFile& delete_file) const {
+    HdfsScanStats app_scan_stats;
+    HdfsScanStats fs_scan_stats;
+    std::shared_ptr<io::SharedBufferedInputStream> shared_buffered_input_stream;
+    std::shared_ptr<io::CacheInputStream> cache_input_stream;
```
Comment (zombee0) [inline be/src/exec/iceberg/iceberg_delete_builder.cpp:80]: how about disabled cache?

File: be/src/exec/iceberg/iceberg_delete_builder.cpp (hunk 1 above)
Comment (stephen-shelby) [reply be/src/exec/iceberg/iceberg_delete_builder.cpp]: done

File: be/src/exec/iceberg/iceberg_delete_builder.cpp (hunk 2 above)
Comment (stephen-shelby) [reply be/src/exec/iceberg/iceberg_delete_builder.cpp]: it's up to config. default is null and I have checked it.

**Review by zombee0 — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Review by trueeyu — APPROVED** [review]

**Pull request merged**